sourcegraph/sourcegraph
# Code insights: Add new repositories UI field

Closes https://github.com/sourcegraph/sourcegraph/issues/45427

## Useful links
- [Original RFC](https://docs.google.com/document/d/1vWxSDSbslZQFEK24rhlo8-K2sCiaaCX5PQoCqEdpI_8/edit#bookmark=id.fch8wwll1mh5)
- [A special doc about repo UI prototype ](https://docs.google.com/document/d/1U_tkYhT1Ai2klXZ9NKllhP_RwXp82cNedaez7rvbnnU/edit#heading=h.2s86rgusc6b3)
- [Figma designs ](https://www.figma.com/file/ZUSD4DTNBdGGomLn6uKZFj/RFC-741-proposal-B%3A-design-for-running-insights-over-partial-sets-%2342333?node-id=246%3A11&t=G7WcsXXF3BYCqPkc-0)

## How to test 
**Locally** 
- Check out this PR branch 
- Run `sg start web-standalone` or any other command that runs FE locally 
- After that you should be able to see new insight creation UI (by default it's two fields for the repo section)
- You can change to the "one field approach" by editing `codeInsightsRepoUI` experimental feature flag in the setting (possible values `"single-search-query", "search-query-or-strict-list"`)

**Remote** 
- Go to the special staging here
- You should see the new repo UI there, you can change UI variation in the same way as it's described in the "Local" section above.

## Backround

This PR changes the repositories UI field for the insight creation UI (search-based and capture group insight creation forms). It adds a new field UI that can be run in two different modes.
- Single repositories search UI _(you can see this mode with `codeInsightsRepoUI: 'single-search-query'` in `experimentalFeatures` setting cascade section)_
<img width="600" alt="Screenshot 2022-12-16 at 17 33 00" src="https://user-images.githubusercontent.com/18492575/208184273-f3cd86e7-0e28-44d4-8cd1-9292e16ae57f.png">

- And repositories search UI + direct list URLs field (the one that we had before this PR) (you can see this mode with `codeInsightsRepoUI: 'search-query-or-strict-list',` in `experimentalFeatures` setting cascade section)
<img width="604" alt="Screenshot 2022-12-16 at 17 32 22" src="https://user-images.githubusercontent.com/18492575/208184270-7d7cb289-80ef-463b-9acf-f70b8a62d44a.png">

Todo list 
- [x] Add setting cascade feature flag
- [x] Implement both modes of UI 
- [x] Add logic for the edit UI (before this PR, we had all repositories mode checkbox, new modes don't have it, in case of all repositories mode is on, we add `repo:.*` to the search repositories query field which should be fully equivalent to the all repos mode that we had before)
- [x] Fix 1-click creation flow for the search-based info (now we just put all repo-related filters to the search repo query field and put the rest of the query from the search result page in the first series query field)
- [x] Connect new UI to the new strat-scoped API handlers - repo query BE validation handler, repositories count info and live preview API 
- [x] Add validation info to the new repositories section in the form (how many repositories were resolved by a given repositories query (search repo query field value) 

## Test plan
- Check that the new UI works and looks as expected
- Check that the 1-click creation flow works as expected
- Check that editing functionality works as it worked before (especially check that all repositories mode has a proper fallback on the `repo:.*` value in the search repo field
- Check that insight templates work as they worked before without any runtime or missing field problems. 




## App preview:

- [Web](https://sg-web-vk-new-insight-repo-ui.onrender.com/search)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.
